feat(lsp): support dynamic registration of didChangeConfiguration

fredrikekre opened the pull request:
Add client support for server to dynamically register the
workspace/didChangeConfiguration handler.

The main motiviation for implementing this feature is that currently
neovim (or rather nvim-lspconfig [0]) sends the didChangeConfiguration
notification just after initialization to pass client settings to the
server regardless whether the server has registered this capability or
not. This is the first step to fix this.

[0]: https://github.com/neovim/nvim-lspconfig/blob/21102d5e3b6ffc6929d60418581ac1a29ee9eddd/lua/lspconfig/configs.lua#L175

fredrikekre committed:
feat(lsp): support dynamic registration of didChangeConfiguration

Add client support for server to dynamically register and unregister the
workspace/didChangeConfiguration handler.

The main motiviation for implementing this feature is that currently
neovim (or rather nvim-lspconfig [0]) sends the didChangeConfiguration
notification just after initialization to pass client settings to the
server regardless whether the server has registered this capability or
not. This is the first step to fix this.

[0]: https://github.com/neovim/nvim-lspconfig/blob/21102d5e3b6ffc6929d60418581ac1a29ee9eddd/lua/lspconfig/configs.lua#L175

mfussenegger reviewed (commented):
The plan is to get rid of the `resolved_capabilities`. See https://github.com/neovim/neovim/pull/17814

And if we implement `registerCapability` for something it will also need the `client/unregisterCapability` part.

fredrikekre commented:
I see, I will revisit this after that PR then.

fredrikekre commented:
Updated the PR after https://github.com/neovim/neovim/pull/17814 and added unregistration. This capability doesn't match directly to something in the servercapability spec, but right now it is stored there anyway (see comments in the code). Will figure out a way to test this too.

justinmk commented:
did https://github.com/neovim/neovim/pull/23681 address this?

mfussenegger commented:
> did https://github.com/neovim/neovim/pull/23681 address this?

It didn't enable the dynamic capability, but it makes this PR in the current form obsolete.

I'm in general not sure if we even should enable `dynamicRegistration = true,` and guard the notification behind a capability check.

There is no static  server capability in the specification for `didChangeConfiguration`, and it's not mentioned if having the `workspace` property can be used to imply it's supported.

I just tested with pylsp, jdtls and luals, and out of the three only luals did set a dynamic `didChangeConfiguration` capability, the others wouldn't have it, despite supporting it.

So I'm closing this. If anyone can point out that the current behavior is causing issues with servers we should ask for clarification in the specification.
